clash-lib/src/Clash/Primitives/Verification.hs

```diff
@@ -0,0 +1,132 @@
+      cvExpr1 <- mapM (uncurry bindMaybe) cvExpr0
+      let decls = concatMap snd cvExpr1
+          cvExpr2 = fmap fst cvExpr1
+      pure (Right (meta, bb (checkTF decls clkId propName renderAs cvExpr2)))
+ where
+  -- TODO: Improve error handling; currently errors don't indicate what
+  -- TODO: blackbox generated them.
+  (Var (id2identifier -> clkId)) = indexNote "clk" (lefts args) 1
+  (Var (id2identifier -> _clkId)) = indexNote "rst" (lefts args) 2
+  propName = uncheckedTermToData (indexNote "propName" (lefts args) 3)
+  renderAs = uncheckedTermToData (indexNote "renderAs" (lefts args) 4)
+  cvProperty = termToDataError (indexNote "propArg" (lefts args) 5)
+
+  bb = BBFunction "Clash.Primitives.Verification.checkTF" 0
+  meta = emptyBlackBoxMeta {bbKind=TDecl, bbRenderVoid=RenderVoid}
+  mkId = Clash.Netlist.Util.mkUniqueIdentifier Basic . Text.pack
+
+  bindMaybe
+    :: Maybe String
+    -- ^ Hint for new identifier
+    -> Term
+    -- ^ Term to bind. Does not bind if it's already a reference to a signal
+    -> NetlistMonad (Identifier, [Declaration])
+    -- ^ ([new] reference to signal, [declarations need to get it in scope])
+  bindMaybe _ (Var vId) = pure (id2identifier vId, [])
+  bindMaybe Nothing t = bindMaybe (Just "s") t
+  bindMaybe (Just nm) t = do
+    tcm <- Lens.use tcCache
+    newId <- mkId nm
+    (expr0, decls) <- mkExpr False Concurrent (NetlistId newId (termType tcm t)) t
+    pure
+      ( newId
+      , decls ++ [sigDecl Bool newId, Assignment newId expr0] )
+
+  -- Simple wire without comment
+  sigDecl :: HWType -> Identifier -> Declaration
+  sigDecl typ nm = NetDecl' Nothing Wire nm (Right typ) Nothing
+
+checkTF
+  :: [Declaration]
+  -> Identifier
+  -> Text.Text
+  -> RenderAs
+  -> Property' Identifier
+  -> TemplateFunction
+checkTF decls clkId propName renderAs prop =
+  TemplateFunction [] (const True) (checkTF' decls clkId propName renderAs prop)
+
+checkTF'
+  :: forall s
+   . (HasCallStack, Backend s)
+  => [Declaration]
+  -- ^ Extra decls needed
+  -> Identifier
+  -- ^ Clock
+  -> Text.Text
+  -- ^ Prop name
+  -> RenderAs
+  -> Property' Identifier
+  -> BlackBoxContext
+  -> State s Doc
+checkTF' decls clkId propName renderAs prop bbCtx = do
+  blockName <- Clash.Backend.mkUniqueIdentifier Basic (propName <> "_block")
+  getMon (blockDecl blockName (TickDecl renderedPslProperty : decls))
+
+ where
+  hdl = hdlKind (undefined :: s)
+
+  edge =
+    case head (bbInputs bbCtx) of
+      (_, stripVoid -> KnownDomain _nm _period e _rst _init _polarity, _) -> e
+      _ -> error $ "Unexpected first argument: " ++ show (head (bbInputs bbCtx))
+
+  renderedPslProperty =
+    let
+      psl = pprPslProperty hdl propName clkId edge prop
+      sva = pprSvaProperty propName clkId edge prop
+    in
+      case renderAs of
+        PSL -> psl
+        SVA -> sva
+        AutoRenderAs ->
+          case hdl of
+            SystemVerilog -> sva
+            Verilog -> psl
+            VHDL -> psl
```

Comment:
What about

```haskell
renderedPslProperty
  | renderAs == SVA || hdl == SystemVerilog = sva
  | otherwise = psl
 where
  sva = pprSvaProperty propName clkId edge prop
  psl = pprPslProperty hdl propName clkId edge prop
```

Reply:
I know it doesn't really look pretty, but I've found ` -Wincomplete-patterns` incredibly useful when changing code I don't know. And in this case I think it's reasonable to assume we're going to add another HDL / verification language at some point.

Reply:
> And in this case I think it's reasonable to assume we're going to add another HDL / verification language at some point.

I'm not so convinced, at least not for the foreseeable future

Reply:
Alright, I'll change it to your version.